### PR TITLE
Add EMS support to tls-tris

### PR DESCRIPTION
**Pull request by wbl**

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -62,6 +62,8 @@ type Conn struct {
 	// renegotiation extension. (This is meaningless as a server because
 	// renegotiation is not supported in that case.)
 	secureRenegotiation bool
+	//useEMS is true if we will use/have used the extended MasterSecret extension
```
Comment (grittygrease) [inline conn.go]: Missing space between // and useEMS.

File: handshake_client.go [hunk 2]
```diff
@@ -423,6 +424,13 @@ func (hs *clientHandshakeState) doFullHandshake() error {
 			return err
 		}
 	}
+	c.useEMS = hs.serverHello.extendedMSSupported
```
Comment (grittygrease) [inline handshake_client.go]: Why was this moved from below the cert chain validation?

File: handshake_messages.go [hunk 3]
```diff
@@ -432,7 +442,7 @@ func (m *clientHelloMsg) unmarshal(data []byte) alert {
 	m.pskKeyExchangeModes = nil
 	m.earlyData = false
 	m.shortHeaders = false
-
```
Comment (grittygrease) [inline handshake_messages.go]: Accidental line removal?

File: handshake_server.go [hunk 4]
```diff
@@ -379,6 +379,11 @@ func (hs *serverHandshakeState) checkForResumption() bool {
 		return false
 	}
 
+	//Upgrade if client now supports EMS
```
Comment (grittygrease) [inline handshake_server.go]: Missing space between // and Upgrade

File: _dev/tris-testclient/client.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+var tlsVersionToName = map[string]uint16{
+	"tls10": tls.VersionTLS10,
+	"tls11": tls.VersionTLS11,
+	"tls12": tls.VersionTLS12,
+	"tls13": tls.VersionTLS13,
+}
+
+//Usage client args host:port
```
Comment (grittygrease) [inline _dev/tris-testclient/client.go]: Missing space between // and Usage

File: handshake_client.go (hunk 2 above)
Comment (wbl) [inline handshake_client.go]: https://tools.ietf.org/html/rfc7627. The handshake hash does not include the client cert chain when the master secret is generated, only the ClientKeyExchange

File: conn.go (hunk 1 above)
Comment (grittygrease) [inline conn.go]: Actually, why is this used for "will"? Shouldn't that setting be part of the Config structure?

**Comment by Lekensteyn** [comment]
(rebased on master, had some conflicts to solve, no other changes)

File: handshake_messages.go [hunk 6]
```diff
@@ -63,6 +64,7 @@ func (m *clientHelloMsg) equal(i interface{}) bool {
 		eqKeyShares(m.keyShares, m1.keyShares) &&
 		eqUint16s(m.supportedVersions, m1.supportedVersions) &&
 		m.earlyData == m1.earlyData
+		m.extendedMSSupported == m1.extendedMSSupported &&
```
Comment (henrydcase) [inline handshake_messages.go]: this won't even compile

File: handshake_messages.go [hunk 7]
```diff
@@ -360,6 +365,11 @@ func (m *clientHelloMsg) marshal() []byte {
 		z[1] = byte(extensionEarlyData)
 		z = z[4:]
```
Comment (henrydcase) [inline handshake_messages.go]: why 4?

File: handshake_messages.go [hunk 8]
```diff
@@ -851,6 +872,11 @@ func (m *serverHelloMsg) marshal() []byte {
 		copy(z, m.secureRenegotiation)
 		z = z[len(m.secureRenegotiation):]
 	}
+	if m.extendedMSSupported {
+		z[0] = byte(extensionEMS >> 8)
+		z[1] = byte(extensionEMS & 0xff)
+		z = z[4:]
```
Comment (henrydcase) [inline handshake_messages.go]: why 4?

File: handshake_messages.go [hunk 9]
```diff
@@ -1029,6 +1052,10 @@ func (m *serverHelloMsg) marshal() []byte {
 		z[5] = byte(m.pskIdentity)
 		z = z[6:]
 	}
+	if m.extendedMSSupported {
+		binary.BigEndian.PutUint16(z, extensionEMS)
+		z = z[4:]
```
Comment (henrydcase) [inline handshake_messages.go]: why 4?

File: handshake_messages.go [hunk 10]
```diff
@@ -425,9 +431,12 @@ func (m *clientHelloMsg) marshal() []byte {
 		z[1] = byte(extensionEarlyData)
 		z = z[4:]
 	}
+	if m.extendedMSSupported {
+		binary.BigEndian.PutUint16(z, extensionEMS)
+		z = z[4:]
```
Comment (henrydcase) [inline handshake_messages.go]: why 4?

File: handshake_messages.go [hunk 11]
```diff
@@ -360,6 +365,11 @@ func (m *clientHelloMsg) marshal() []byte {
 		z[1] = byte(extensionEarlyData)
 		z = z[4:]
 	}
+	if m.extendedMSSupported {
+		z[0] = byte(extensionEMS >> 8)
+		z[1] = byte(extensionEMS & 0xff)
+		z = z[4:]
```
[the remark this answers is not in the document]
Comment (cjpatton) [reply handshake_messages.go]: 2 bytes for the extension code point, 2 bytes for the length of the extension. (Should be 0 for subcerts, but others actually encode a length.)

[the remark this answers is not in the document]
Comment (henrydcase) [reply handshake_messages.go]: yep, my point was - "why hardcoding 4" and not a constant which makes it easy to read the code...

File: handshake_messages_test.go [hunk 12]
```diff
@@ -172,6 +172,8 @@ func (*clientHelloMsg) Generate(rand *rand.Rand, size int) reflect.Value {
 		m.earlyData = true
 	}
 
+	m.delegatedCredential = true
```
Comment (henrydcase) [inline handshake_messages_test.go]: So, this is wrong - delegatedCredentials and EMS should be set randomly to true/false

File: conn.go [hunk 13]
```diff
@@ -66,6 +66,8 @@ type Conn struct {
 	// renegotiation extension. (This is meaningless as a server because
 	// renegotiation is not supported in that case.)
 	secureRenegotiation bool
+	// useEMS is true if we have used the extended MasterSecret extension
```
Comment (henrydcase) [inline conn.go]: improve comment

File: handshake_client.go [hunk 14]
```diff
@@ -697,6 +699,16 @@ func (hs *clientHandshakeState) processServerHello() (bool, error) {
 		}
 	}
 
+	if hs.serverHello.extendedMSSupported {
+		if hs.hello.extendedMSSupported {
+			c.useEMS = true
+		} else {
+			// TODO(kk) It doesn't seem to be in line with RFC7627
```
Comment (henrydcase) [inline handshake_client.go]: Needs fixing

File: prf.go [hunk 15]
```diff
@@ -141,14 +142,21 @@ func prfForVersion(version uint16, suite *cipherSuite) func(result, secret, labe
...
 // secret. See http://tools.ietf.org/html/rfc5246#section-8.1
-func masterFromPreMasterSecret(version uint16, suite *cipherSuite, preMasterSecret, clientRandom, serverRandom []byte) []byte {
-	seed := make([]byte, 0, len(clientRandom)+len(serverRandom))
-	seed = append(seed, clientRandom...)
-	seed = append(seed, serverRandom...)
-
-	masterSecret := make([]byte, masterSecretLength)
-	prfForVersion(version, suite)(masterSecret, preMasterSecret, masterSecretLabel, seed)
-	return masterSecret
+func masterFromPreMasterSecret(version uint16, suite *cipherSuite, preMasterSecret, clientRandom, serverRandom []byte, fin finishedHash, ems bool) []byte {
```
Comment (henrydcase) [inline prf.go]: I don't like this function.
* clientRandom/serverRandom are used only for !ems
* the 'fin' thing is only used for ems

*
